ships/0025-event-driven-builds.md

```diff
@@ -0,0 +1,300 @@
+   expected by the `Run` resource ie, the branch name & the commit ID.
+   
+ ```
+ 
+### Generated by the `Build` reconciler. 
+ 
+apiVersion: triggers.tekton.dev/v1alpha1
+kind: TriggerTemplate
+metadata:
+ name: my-build-name
+spec:
+ params:
+   - name: branch
+   - name: commit
+ resourceTemplates:
+   - apiVersion: tekton.dev/v1alpha1
+     kind: Run
+     metadata:
+       generateName: build-execution-
+     spec:
+       ref:
+         apiVersion: shipwright.io/v1alpha1
+         kind: Build 
+         name: my-build-bame
+       timeout: 3000s
+       params:
+         - name: branch
+           value: $(tt.params.branch)
+         - name: commit
+           value: $(tt.params.commit)
+ ```
+
+2. The Shipwright Build Controller would create a `EventListener` under-the-hood per build. 
+    
+```
+### Generated by the `Build` reconciler. 
+
+apiVersion: triggers.tekton.dev/v1alpha1
+kind: EventListener
+metadata:
+ name: name-of-the-build
+spec:
+ serviceAccountName: pipeline
+ triggers:
+   - bindings:
+       - ref: github-shipwright-webhook
+     template:
+       name: name-of-the-build
+   
+```
+
+And that's it, you may now go ahead and 
+
+### Test Plan
+
+To be filled.
+
+### Release Criteria
+
+To be filled.
+
+#### Removing a deprecated feature [if necessary]
+
+N/A
+
+#### Upgrade Strategy [if necessary]
+
+N/A
+
+### Risks and Mitigations
+
+1. This feature opens up the possiblility of triggering Build executions for branches which weren't explicitly specified. 
+
+This isn't a risk per se since only repository committers would have the permissions needed to push branches. Therefore,
+as long as the image is tagged appropriately to indiciate that it is not built off 'main' ( or the branch specified iniially by the user in the `Build`), 
+this should not be a problem.
+
+While the enhancement could have been scoped to only deal with the branch/revision explicitly specified in the `Build` resource, 
+doing the same would have have excluded some key use cases where users push to a different branch before merging to 'main'. 
+
+2. Exposing a webhook URL enables creation of pods ( ie, processes on the node ) by actors who may not necessarily have access to the cluster.
+
+This does open up an attack vector given the execution of the build is done using the configured service account. 
+
+* Webhook-driven builds are being designed to be secure by default - the usage of a webhook secret is mandatory.
+* The resulting `BuildRun` would be annotated with relevant metadata from the webhook event so that it's easy to trace the actor responsible for the trigger. 
+
+
+## Drawbacks
+
+1. Is this even a Shipwright concern or should this be the concern of a general-purpose CI ?
+( to be filled ) 
+
+
+
+
+## Alternatives
+
+Similar to the `Drawbacks` section the `Alternatives` section is used to highlight and record other
+possible approaches to delivering the value proposed by an enhancement.
```

Comment:
How much simpler would it be if we implement our own webhook endpoint that directly consumes the event data from GitHub or GitLab? I personally think, it is much simpler. And it will be trivial to implement logic there that checks the branch of the payload and compares it with what is defined in the Build.

Reply:
You mean -  write a simple http service which understands the payload and creates corresponding Build/BuildRuns?

Reply:
Yes.